instrumentation/spring/spring-cloud-gateway/spring-cloud-gateway-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/spring/gateway/v2_0/ServerWebExchangeHelper.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.spring.gateway.v2_0;
+
+import static org.springframework.cloud.gateway.support.ServerWebExchangeUtils.GATEWAY_ROUTE_ATTR;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.internal.StringUtils;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.javaagent.bootstrap.internal.InstrumentationConfig;
+import java.util.regex.Pattern;
+import org.springframework.cloud.gateway.route.Route;
+import org.springframework.web.server.ServerWebExchange;
+
+public final class ServerWebExchangeHelper {
+
+  /** Route info key. */
+  private static final AttributeKey<String> ROUTE_INFO_ATTRIBUTES =
+      AttributeKey.stringKey("ROUTE_INFO");
+
+  private static final boolean CAPTURE_EXPERIMENTAL_SPAN_ATTRIBUTES;
+
+  static {
+    CAPTURE_EXPERIMENTAL_SPAN_ATTRIBUTES =
+        InstrumentationConfig.get()
+            .getBoolean(
+                "otel.instrumentation.spring-cloud-gateway.experimental-span-attributes", false);
+  }
+
+  /* Regex for UUID */
+  private static final Pattern UUID_REGEX =
+      Pattern.compile(
+          "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$");
+
+  /* Route ID for route whose routeID is unset */
+  private static final String UNSET_ROUTE_ID = "UNSET_ROUTE_ID";
+
+  private static final String INVALID_RANDOM_ROUTE_ID =
+      "org.springframework.util.AlternativeJdkIdGenerator@";
+
+  private ServerWebExchangeHelper() {}
+
+  public static void extractAttributes(ServerWebExchange exchange, Context context) {
+    // Record route info
+    Route route = exchange.getAttribute(GATEWAY_ROUTE_ATTR);
+    if (route != null && CAPTURE_EXPERIMENTAL_SPAN_ATTRIBUTES) {
+      Span currentSpan = Span.fromContext(context);
+      if (currentSpan == null) {
+        return;
+      }
+      currentSpan.setAttribute(ROUTE_INFO_ATTRIBUTES, summarizeRoute(route));
+    }
+  }
+
+  public static String extractServerRoute(ServerWebExchange exchange) {
+    Route route = exchange.getAttribute(GATEWAY_ROUTE_ATTR);
+    if (route != null) {
+      return convergeRouteId(route);
+    }
+    return null;
+  }
+
+  /**
+   * To avoid high cardinality, we ignore random UUID generated by Spring Cloud Gateway. Spring
+   * Cloud Gateway generate invalid random routeID, and it is fixed until 3.1.x
+   *
+   * @see <a
+   *     href="https://github.com/spring-cloud/spring-cloud-gateway/commit/5002fe2e0a2825ef47dd667cade37b844c276cf6"/>
+   */
+  private static String convergeRouteId(Route route) {
+    String routeId = route.getId();
+    if (StringUtils.isNullOrEmpty(routeId)) {
+      return UNSET_ROUTE_ID;
```

Comment:
If a meaningful route can not be extracted I'd just return null and let it use the default span name.